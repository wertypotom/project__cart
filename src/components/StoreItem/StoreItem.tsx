import React, { useContext } from 'react';
import { Button, Card } from 'react-bootstrap';
import cartContext from '../../context/cartContext';
import formatCurrency from '../../utilities/currencyFormater';

type Props = {
  id: number;
  name: string;
  imgUrl: string;
  price: number;
  amount: number;
};

function StoreItem({ id, imgUrl, name, price, amount }: Props) {
  const { toggleItem } = useContext(cartContext);

  return (
    <Card className="h-100">
      <Card.Img
        variant="top"
        src={imgUrl}
        height="200px"
        style={{ objectFit: 'cover' }}
      />
      <Card.Body className="d-flex flex-column">
        <Card.Title className="d-flex justify-content-between align-items-baseline mb-4">
          <span className="fs-2">{name}</span>
          <span className="ms-2 text-muted">{formatCurrency(price)}</span>
        </Card.Title>
        <div className="mt-auto">
          {amount === 0 ? (
            <Button
              data-testid="add-to-cart"
              className="w-100"
              onClick={() => toggleItem(id, 'add')}
            >
              Add to Cart
            </Button>
          ) : (
            <div
              data-testid="toggle-btns"
              className="d-flex align-items-center flex-column"
              style={{ gap: '0.5rem' }}
            >
              <div>
                <Button onClick={() => toggleItem(id, 'remove')}>-</Button>
                <span className="mx-2">{amount} in cart</span>
                <Button onClick={() => toggleItem(id, 'add')}>+</Button>
              </div>
              <Button
                variant="danger"
                size="sm"
                onClick={() => toggleItem(id, 'reset')}
              >
                Remove
              </Button>
            </div>
          )}
        </div>
      </Card.Body>
    </Card>
  );
}

export default StoreItem;
