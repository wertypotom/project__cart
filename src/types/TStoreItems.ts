export type TStoreItems = {
  id: number;
  name: string;
  price: number;
  imgUrl: string;
  amount: number;
};

export type TItemToggle = 'add' | 'remove' | 'reset';
