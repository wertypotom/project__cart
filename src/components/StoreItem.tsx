import React from 'react';
import { Button, Card } from 'react-bootstrap';
import { TItemToggle } from '../types/TStoreItems';
import formatCurrency from '../utilities/currencyFormater';

type Props = {
  id: number;
  name: string;
  imgUrl: string;
  price: number;
  amount: number;
  toggleItem: (id: number, type: TItemToggle) => void;
};

function StoreItem({
  id, imgUrl, name, price, amount, toggleItem,
}: Props) {
  return (
    <Card className="h-100">
      <Card.Img
        variant="top"
        src={imgUrl}
        height="200px"
        style={{ objectFit: 'cover' }}
      />
      <Card.Body className="d-flex flex-column">
        <Card.Title className="d-flex justify-content-between align-items-baseline mb-4">
          <span className="fs-2">{name}</span>
          <span className="ms-2 text-muted">{formatCurrency(price)}</span>
        </Card.Title>
        <div className="mt-auto">
          {amount === 0 ? (
            <Button className="w-100" onClick={() => toggleItem(id, 'add')}>
              Add to Cart
            </Button>
          ) : (
            <div
              className="d-flex align-items-center flex-column"
              style={{ gap: '0.5rem' }}
            >
              <div>
                <Button onClick={() => toggleItem(id, 'remove')}>-</Button>
                <span className="mx-2">
                  {amount}
                  {' '}
                  in cart
                </span>
                <Button onClick={() => toggleItem(id, 'add')}>+</Button>
              </div>
              <Button
                variant="danger"
                size="sm"
                onClick={() => toggleItem(id, 'reset')}
              >
                Remove
              </Button>
            </div>
          )}
        </div>
      </Card.Body>
    </Card>
  );
}

export default StoreItem;
