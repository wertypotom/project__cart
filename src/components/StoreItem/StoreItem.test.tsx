import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import CartContext from '../../context/cartContext';
import { TStoreItems } from '../../types/TStoreItems';
import StoreItem from './StoreItem';

describe('Test StoreItem Component', () => {
  let item: TStoreItems;
  const items: TStoreItems[] = [
    {
      id: 1,
      name: 'Book',
      price: 10.99,
      imgUrl: '/imgs/book.jpg',
      amount: 0,
    },
    {
      id: 2,
      name: 'Table',
      price: 10.99,
      imgUrl: '/imgs/book.jpg',
      amount: 0,
    },
    {
      id: 3,
      name: 'Toy',
      price: 10.99,
      imgUrl: '/imgs/book.jpg',
      amount: 0,
    },
  ];
  const toggleItem = vi.fn(() => {});
  const { getByTestId, queryByTestId, getByText } = screen;

  beforeEach(() => {
    item = {
      id: 1,
      name: 'Book',
      price: 10.99,
      imgUrl: '/imgs/book.jpg',
      amount: 0,
    };
  });

  it('should create snapshot', () => {
    const componenet = render(<StoreItem {...item} />);
    expect(componenet).toMatchSnapshot();
  });

  it('should render "Add to cart" Btn', () => {
    const componenet = render(
      <CartContext.Provider value={{ toggleItem, items }}>
        <StoreItem {...item} />
      </CartContext.Provider>
    );

    const btn = getByTestId('add-to-cart');
    expect(btn).toBeInTheDocument();

    screen.debug();

    // const itemsAmount = getByText('1 in cart');
    // expect(itemsAmount).toBeInTheDocument();
    // const toggleBtns = getByTestId('toggle-btns');
    // expect(toggleBtns).toBeInTheDocument();
  });

  it('should render toggle btns', () => {
    item.amount = 2;
    const componenet = render(<StoreItem {...item} />);

    const addToCartBtn = queryByTestId('add-to-cart');
    const toggleBtns = getByTestId('toggle-btns');
    expect(addToCartBtn).not.toBeInTheDocument();
    expect(toggleBtns).toBeInTheDocument();
  });
});
