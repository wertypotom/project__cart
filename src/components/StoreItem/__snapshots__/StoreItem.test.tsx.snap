// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test StoreItem Component > should create snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-100 card"
      >
        <img
          class="card-img-top"
          height="200px"
          src="/imgs/book.jpg"
          style="object-fit: cover;"
        />
        <div
          class="d-flex flex-column card-body"
        >
          <div
            class="d-flex justify-content-between align-items-baseline mb-4 card-title h5"
          >
            <span
              class="fs-2"
            >
              Book
            </span>
            <span
              class="ms-2 text-muted"
            >
              10,99 $
            </span>
          </div>
          <div
            class="mt-auto"
          >
            <button
              class="w-100 btn btn-primary"
              data-testid="add-to-cart"
              type="button"
            >
              Add to Cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-100 card"
    >
      <img
        class="card-img-top"
        height="200px"
        src="/imgs/book.jpg"
        style="object-fit: cover;"
      />
      <div
        class="d-flex flex-column card-body"
      >
        <div
          class="d-flex justify-content-between align-items-baseline mb-4 card-title h5"
        >
          <span
            class="fs-2"
          >
            Book
          </span>
          <span
            class="ms-2 text-muted"
          >
            10,99 $
          </span>
        </div>
        <div
          class="mt-auto"
        >
          <button
            class="w-100 btn btn-primary"
            data-testid="add-to-cart"
            type="button"
          >
            Add to Cart
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
